### PR TITLE
unravel_index check if all elements in `indices` are within the acceptable range

### LINKED ISSUE #136736: torch.unravel_index does not check out-of-bounds
### 🐛 Describe the bug

I'm unsure if this is intended behaviour or not, but when using torch.unravel_index and entering indices that are out-of-bounds, the returned indices wrap back around to the start of the tensor.

For example, unvravelling indices 0 through 8 for a 2x2 array raises an error in numpy:
`np.unravel_index(np.arange(0,9,1), (2,2))`
`ValueError: index 4 is out of bounds for array with size 4`

but returns two copies of every index in torch:
`torch.unravel_index(torch.arange(0,9,1), (2,2))`
`(tensor([0, 0, 1, 1, 0, 0, 1, 1, 0]), tensor([0, 1, 0, 1, 0, 1, 0, 1, 0]))`

### Versions

PyTorch version: 2.3.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.6.1 (x86_64)
GCC version: Could not collect
Clang version: 12.0.5 (clang-1205.0.22.9)
CMake version: version 3.24.4
Libc version: N/A

Python version: 3.12.4 | packaged by Anaconda, Inc. | (main, Jun 18 2024, 10:14:12) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-10.16-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] torch==2.3.0
[pip3] torchviz==0.0.2
[conda] numpy                     1.26.4          py312h7d6adbd_0  
[conda] numpy-base                1.26.4          py312hbb3573c_0  
[conda] pytorch                   2.3.0           cpu_py312h9d484b6_0  
[conda] torchviz                  0.0.2                    pypi_0    pypi

### PR DESCRIPTION
**Pull request by Rahul-Ram-03**
Fixes #136736 

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @mcarilli @ptrblck @leslie-fang-intel

### CODE REVIEW

**Comment by Rahul-Ram-03** [comment]
Just noticed the CC spam. I am not entirely sure how that happened; I am a new contributor so, please forgive me.

File: torch/functional.py [hunk 1]
```diff
@@ -1988,6 +1988,12 @@ def _unravel_index(indices: Tensor, shape: Union[int, Sequence[int]]) -> Tensor:
             )
         )
     )
+
+    torch._check_value(
+        all(index < coefs[0] * shape[0] for index in indices),
+        lambda: f"'indices' must be in range {[0, coefs[0] * shape[0] - 1]}, but got {indices} instead.",
+    )
```
Comment (nikitaved) [inline torch/functional.py:1995]: This is exactly why we avoid checking indices. The problem is that if tensors live on the GPU, the `all` will cause a CPU-GPU synchronization, which can be expensive. Is the check missing in the kernel? If so, it is best to be placed there.

**Review by ezyang — CHANGES_REQUESTED** [review]
This implementation triggers DtoH sync and cannot be landed as is

**Comment by Rahul-Ram-03** [comment]
> This implementation triggers DtoH sync and cannot be landed as is

I assume the issue is with the fact that `coefs` is a "local" variable and as such, exists on the host, whilst `indices` could potentially exist on device. And I also assume that using `coefs[0]` causes the entire array to move to the device. As a workaround, should I initialize a tensor with just the value of `coefs[0]`?

**Comment by ezyang** [comment]
The problem is not coefs, it's indices, which is a Tensor

File: torch/functional.py (hunk 1 above)
Comment (eellison) [inline torch/functional.py:1995]: can you do assert_async instead ?

Comment (Rahul-Ram-03) [reply torch/functional.py]: Thank you for the pointer!

**Comment by Rahul-Ram-03** [comment]
Paging @eellison @ezyang

**Review by ezyang — CHANGES_REQUESTED** [review]
Nope

**Pull request closed without merging**